File: sdk/cpp/tests/test_metric.cpp
/**
 * Tests for the0::metric() and the0::log() functions
 *
 * These functions write to stdout (not to result file).
 */

#include <doctest/doctest.h>
#include "../the0.h"
#include "helpers/test_helpers.hpp"

using namespace the0_test;

TEST_SUITE("the0::metric") {
    TEST_CASE("outputs JSON with _metric field to stdout") {
        CaptureStdout capture;
        the0::metric("price", {{"symbol", "BTC/USD"}, {"value", 45000.0}});

        std::string output = capture.get();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["_metric"] == "price");
        CHECK(json["symbol"] == "BTC/USD");
        CHECK(json["value"] == 45000.0);
    }

    TEST_CASE("includes timestamp field") {
        CaptureStdout capture;
        the0::metric("signal", {{"type", "BUY"}});

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json.contains("timestamp"));
        CHECK(json["timestamp"].is_string());
        // Timestamp should end with 'Z'
        std::string ts = json["timestamp"].get<std::string>();
        CHECK(ts.back() == 'Z');
    }

    TEST_CASE("handles empty data") {
        CaptureStdout capture;
        the0::metric("heartbeat", {});

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["_metric"] == "heartbeat");
        CHECK(json.contains("timestamp"));
    }

    TEST_CASE("handles nested data") {
        CaptureStdout capture;
        the0::metric("trade", {
            {"order", {
                {"id", "12345"},
                {"side", "buy"},
                {"quantity", 1.5}
            }}
        });

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["_metric"] == "trade");
        CHECK(json["order"]["id"] == "12345");
        CHECK(json["order"]["side"] == "buy");
        CHECK(json["order"]["quantity"] == 1.5);
    }

    TEST_CASE("handles array data") {
        CaptureStdout capture;
        the0::metric("prices", {{"values", {100.0, 101.5, 99.8}}});

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["values"].is_array());
        CHECK(json["values"].size() == 3);
        CHECK(json["values"][0] == 100.0);
    }

    TEST_CASE("outputs newline at end") {
        CaptureStdout capture;
        the0::metric("test", {});

        std::string output = capture.get();
        CHECK(output.back() == '\n');
    }
}

TEST_SUITE("the0::log") {
    TEST_CASE("outputs JSON with message field to stderr") {
        CaptureStderr capture;
        the0::log("Starting bot execution");

        std::string output = capture.get();
        REQUIRE(!output.empty());

        auto json = the0::json::parse(output);
        CHECK(json["message"] == "Starting bot execution");
    }

    TEST_CASE("defaults to info level") {
        CaptureStderr capture;
        the0::log("Test message");

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["level"] == "info");
    }

    TEST_CASE("includes timestamp") {
        CaptureStderr capture;
        the0::log("Test message");

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json.contains("timestamp"));
        std::string ts = json["timestamp"].get<std::string>();
        CHECK(ts.back() == 'Z');
    }

    TEST_CASE("supports warn level") {
        CaptureStderr capture;
        the0::log("Warning message", {}, the0::LogLevel::Warn);

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["level"] == "warn");
        CHECK(json["message"] == "Warning message");
    }

    TEST_CASE("supports error level") {
        CaptureStderr capture;
        the0::log("Error message", {}, the0::LogLevel::Error);

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["level"] == "error");
        CHECK(json["message"] == "Error message");
    }

    TEST_CASE("merges data fields") {
        CaptureStderr capture;
        the0::log("Order placed", {{"order_id", "12345"}, {"symbol", "BTC"}});

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["message"] == "Order placed");
        CHECK(json["order_id"] == "12345");
        CHECK(json["symbol"] == "BTC");
    }

    TEST_CASE("supports data and level together") {
        CaptureStderr capture;
        the0::log("Order failed", {{"order_id", "12345"}}, the0::LogLevel::Error);

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["level"] == "error");
        CHECK(json["message"] == "Order failed");
        CHECK(json["order_id"] == "12345");
    }

    TEST_CASE("properly escapes special characters in message") {
        CaptureStderr capture;
        the0::log("Error: \"file not found\"");

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["message"] == "Error: \"file not found\"");
    }

    TEST_CASE("handles empty message") {
        CaptureStderr capture;
        the0::log("");

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["message"] == "");
    }

    TEST_CASE("handles newlines in message") {
        CaptureStderr capture;
        the0::log("Line 1\nLine 2");

        std::string output = capture.get();
        auto json = the0::json::parse(output);

        CHECK(json["message"] == "Line 1\nLine 2");
    }

    TEST_CASE("outputs newline at end") {
        CaptureStderr capture;
        the0::log("test");

        std::string output = capture.get();
        CHECK(output.back() == '\n');
    }
}

File: sdk/rust/tests/test_metric.rs
//! Tests for the0_sdk::input::metric() and log() functions
//!
//! These functions write to stdout. Since capturing stdout in Rust tests
//! is complex, we primarily test that they don't panic and produce valid output.
//! Full integration testing is done via the example bots.

mod common;

use common::EnvGuard;
use serde_json::json;
use the0_sdk::input;

#[test]
fn metric_does_not_panic_with_valid_data() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "metric-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::metric("price", &json!({"symbol": "BTC/USD", "value": 45000.0}));
}

#[test]
fn metric_does_not_panic_with_empty_data() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "metric-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::metric("heartbeat", &json!({}));
}

#[test]
fn metric_does_not_panic_with_nested_data() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "metric-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::metric("trade", &json!({
        "order": {
            "id": "12345",
            "side": "buy",
            "quantity": 1.5
        }
    }));
}

#[test]
fn metric_does_not_panic_with_array_data() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "metric-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::metric("prices", &json!({"values": [100.0, 101.5, 99.8]}));
}

#[test]
fn log_does_not_panic_with_message() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic - simple log with defaults
    input::log("Starting bot execution", None, None);
}

#[test]
fn log_does_not_panic_with_level() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic - log with level
    input::log("Warning message", None, Some(input::LogLevel::Warn));
    input::log("Error message", None, Some(input::LogLevel::Error));
}

#[test]
fn log_does_not_panic_with_data() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic - log with structured data
    input::log("Order placed", Some(&json!({"order_id": "123", "symbol": "BTC"})), None);
}

#[test]
fn log_does_not_panic_with_data_and_level() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic - log with data and level
    input::log("Order failed", Some(&json!({"order_id": "123"})), Some(input::LogLevel::Error));
}

#[test]
fn log_does_not_panic_with_empty_message() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::log("", None, None);
}

#[test]
fn log_does_not_panic_with_special_characters() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic - quotes and backslashes get escaped
    input::log(r#"Error: "file not found" at C:\path"#, None, None);
}

#[test]
fn log_does_not_panic_with_newlines() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::log("Line 1\nLine 2\nLine 3", None, None);
}

#[test]
fn log_does_not_panic_with_long_message() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    let long_message = "x".repeat(10000);
    // Should not panic
    input::log(&long_message, None, None);
}

#[test]
fn log_convenience_functions_do_not_panic() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "log-test");
    env.set("BOT_CONFIG", "{}");

    // Test convenience functions
    input::log_info("Info message", None);
    input::log_warn("Warning message", None);
    input::log_error("Error message", None);
    input::log_info("With data", Some(&json!({"key": "value"})));
}

#[test]
fn metric_handles_various_numeric_types() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "numeric-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic with various numeric types
    input::metric("numbers", &json!({
        "integer": 42,
        "negative": -100,
        "float": 3.14159,
        "large": 9999999999_i64,
        "small": 0.0000001
    }));
}

#[test]
fn metric_handles_boolean_and_null() {
    let mut env = EnvGuard::new();
    env.set("BOT_ID", "types-test");
    env.set("BOT_CONFIG", "{}");

    // Should not panic
    input::metric("types", &json!({
        "active": true,
        "disabled": false,
        "value": null
    }));
}
